### DISCUSSION TITLE
Can parse a 70 line file, but not a line > 70 file (using Rust parser)

### DISCUSSION THREAD

**Discussion by MostHated** [tree-sitter/tree-sitter#2228] [Q&A] [answered]
Hey there, 
I am not sure what is going on here. I am not getting any sort of parse errors, nothing weird shows up in the syntax tree, but if I try to parse a file that is 71 lines or more and I do a debug pring of the Tree, it simply returns the root node with no children or rows/cols:

```rust
// Parse via the following:
let tree = &parser.parse(content, None);

// debug printing the tree
Syntax Tree: Child Count: 0
source_file [0, 0] - [0, 0]
```

But then if I remove a line from the file (doesn't seemingly matter what line), the parse function returns the proper tree. 

<details>
 <summary> Tree when file is 70 lines </summary>

```js
Syntax Tree: Child Count: 19
source_file [0, 0] - [70, 0] 
  import_statement [0, 0] - [0, 15] 
    identifier [0, 7] - [0, 8] 
    string [0, 9] - [0, 15] 
  import_statement [1, 0] - [1, 14] 
    identifier [1, 7] - [1, 9] 
    string [1, 10] - [1, 14] 
  comment [3, 0] - [3, 10] 
  comment [4, 0] - [4, 20] 
  comment [6, 0] - [6, 32] 
  try_statement [7, 0] - [13, 2] 
    try_identifier [7, 0] - [7, 3] 
    block [7, 4] - [8, 18] 
      throw_expression [8, 2] - [8, 18] 
        throw_identifier [8, 2] - [8, 7] 
        error_expression [8, 8] - [8, 18] 
          error [8, 8] - [8, 13] 
          identifier [8, 14] - [8, 18] 
            var_identifier [8, 14] - [8, 18] 
    catch_statement [9, 0] - [13, 2] 
      catch_identifier [9, 0] - [9, 5] 
      catch_block [9, 6] - [13, 2] 
        field_expression [10, 2] - [13, 2] 
          identifier [10, 2] - [10, 3] 
          call_expression [10, 4] - [13, 2] 
            identifier [10, 4] - [10, 6] 
            parameter [10, 7] - [10, 20] 
              call_expression [10, 7] - [10, 20] 
                identifier [10, 7] - [10, 18] 
            parameter [10, 22] - [13, 1] 
              multiline_string [10, 22] - [13, 1] 
  comment [15, 0] - [15, 55] 
  function_definition [16, 0] - [17, 18] 
    func [16, 0] - [16, 4] 
    function_declaration [16, 5] - [16, 11] 
      identifier [16, 5] - [16, 9] 
    block [16, 12] - [17, 18] 
      throw_expression [17, 2] - [17, 18] 
        throw_identifier [17, 2] - [17, 7] 
        error_expression [17, 8] - [17, 18] 
          error [17, 8] - [17, 13] 
          identifier [17, 14] - [17, 18] 
            var_identifier [17, 14] - [17, 18] 
  function_definition [18, 0] - [31, 3] 
    func [18, 0] - [18, 4] 
    function_declaration [18, 5] - [18, 10] 
      identifier [18, 5] - [18, 8] 
    block [18, 11] - [31, 3] 
      try_statement [19, 2] - [31, 3] 
        try_identifier [19, 2] - [19, 5] 
        block [19, 6] - [20, 10] 
          call_expression [20, 4] - [20, 10] 
            identifier [20, 4] - [20, 8] 
        catch_statement [21, 2] - [31, 3] 
          catch_identifier [21, 2] - [21, 7] 
          catch_block [21, 8] - [31, 3] 
            field_expression [22, 4] - [31, 3] 
              identifier [22, 4] - [22, 5] 
              call_expression [22, 6] - [31, 3] 
                identifier [22, 6] - [22, 8] 
                parameter [22, 9] - [22, 22] 
                  call_expression [22, 9] - [22, 22] 
                    identifier [22, 9] - [22, 20] 
                parameter [22, 24] - [31, 1] 
                  multiline_string [22, 24] - [31, 1] 
  call_expression [32, 0] - [32, 5] 
    identifier [32, 0] - [32, 3] 
  local_declaration [34, 0] - [37, 1] 
    identifier [34, 0] - [34, 3] 
    multiline_string [34, 6] - [37, 1] 
  comment [38, 0] - [38, 10] 
  comment [39, 0] - [39, 20] 
  comment [40, 0] - [40, 32] 
  try_statement [41, 0] - [47, 2] 
    try_identifier [41, 0] - [41, 3] 
    block [41, 4] - [42, 18] 
      throw_expression [42, 2] - [42, 18] 
        throw_identifier [42, 2] - [42, 7] 
        error_expression [42, 8] - [42, 18] 
          error [42, 8] - [42, 13] 
          identifier [42, 14] - [42, 18] 
            var_identifier [42, 14] - [42, 18] 
    catch_statement [43, 0] - [47, 2] 
      catch_identifier [43, 0] - [43, 5] 
      catch_block [43, 6] - [47, 2] 
        field_expression [44, 2] - [47, 2] 
          identifier [44, 2] - [44, 3] 
          call_expression [44, 4] - [47, 2] 
            identifier [44, 4] - [44, 6] 
            parameter [44, 7] - [44, 20] 
              call_expression [44, 7] - [44, 20] 
                identifier [44, 7] - [44, 18] 
            parameter [44, 22] - [47, 1] 
              multiline_string [44, 22] - [47, 1] 
  try_statement [49, 0] - [52, 41] 
    try_identifier [49, 0] - [49, 3] 
    block [49, 4] - [50, 18] 
      throw_expression [50, 2] - [50, 18] 
        throw_identifier [50, 2] - [50, 7] 
        error_expression [50, 8] - [50, 18] 
          error [50, 8] - [50, 13] 
          identifier [50, 14] - [50, 18] 
            var_identifier [50, 14] - [50, 18] 
    catch_statement [51, 0] - [52, 41] 
      catch_identifier [51, 0] - [51, 5] 
      catch_block [51, 6] - [52, 41] 
        field_expression [52, 2] - [52, 41] 
          identifier [52, 2] - [52, 3] 
          call_expression [52, 4] - [52, 41] 
            identifier [52, 4] - [52, 6] 
            parameter [52, 7] - [52, 20] 
              call_expression [52, 7] - [52, 20] 
                identifier [52, 7] - [52, 18] 
            parameter [52, 22] - [52, 40] 
              multiline_string [52, 22] - [52, 40] 
  comment [55, 0] - [55, 55] 
  function_definition [56, 0] - [57, 18] 
    func [56, 0] - [56, 4] 
    function_declaration [56, 5] - [56, 11] 
      identifier [56, 5] - [56, 9] 
    block [56, 12] - [57, 18] 
      throw_expression [57, 2] - [57, 18] 
        throw_identifier [57, 2] - [57, 7] 
        error_expression [57, 8] - [57, 18] 
          error [57, 8] - [57, 13] 
          identifier [57, 14] - [57, 18] 
            var_identifier [57, 14] - [57, 18] 
  function_definition [58, 0] - [69, 2] 
    func [58, 0] - [58, 4] 
    function_declaration [58, 5] - [58, 10] 
      identifier [58, 5] - [58, 8] 
    block [58, 11] - [69, 2] 
      try_statement [59, 2] - [69, 2] 
        try_identifier [59, 2] - [59, 5] 
        block [59, 6] - [60, 10] 
          call_expression [60, 4] - [60, 10] 
            identifier [60, 4] - [60, 8] 
        catch_statement [61, 2] - [69, 2] 
          catch_identifier [61, 2] - [61, 7] 
          catch_block [61, 8] - [69, 2] 
            field_expression [62, 4] - [69, 2] 
              identifier [62, 4] - [62, 5] 
              call_expression [62, 6] - [69, 2] 
                identifier [62, 6] - [62, 8] 
                parameter [62, 9] - [62, 22] 
                  call_expression [62, 9] - [62, 22] 
                    identifier [62, 9] - [62, 20] 
                parameter [62, 24] - [69, 1] 
                  multiline_string [62, 24] - [69, 1]
```

</details>

Output at lines > 70 

The odd thing is, it seems to work fine regardless when doing:

```sh
tree-sitter parse filename
```

Though, just in case, here is the output of the above command with ` --debug `

https://gist.github.com/MostHated/15fb2872172e3f0755679021adc5c7bd

If anyone happens to have any ideas, I would sure appreciate hearing them.
Thanks,
-MH

**Comment by ahlinc** [comment]
Tried and the [Rust](https://github.com/tree-sitter/tree-sitter-rust) parser works fine on the master.
Please provide a full input sample and try to use `tree-sitter` CLI downloaded from the [releases](https://github.com/tree-sitter/tree-sitter/releases) instead of installed from npm.

**Comment by MostHated** [comment]
My apologies, I should have clarified. It was not using the parser for the rust language, [it was a parser I have been working on](https://github.com/instance-id/tree-sitter-cyber), being used in the rust language via bindings.

This was the [cyber language](https://cyberscript.dev) script in paricular I was using to test. 
https://gist.github.com/MostHated/657a6132c998578f60d6634dfdb1a610

[This is the project](https://github.com/instance-id/cyber-language-server/blob/903e5e5a0b3b3a73be5b6fae4b59d08e61e7059d/crates/cyber-lsp/src/handlers.rs#L202) I am building in which I am using the parser, pointing to where I am acually parsing a file using it.

I am about to test again following your recommendation to use a build from here. I will report back shortly.

----

Edit: I tried building the rust bindings again with the v20.8 binary, but unfortunately, I am seeing the same behavior.

Not super helpful, but here is a demonstration. I start out with the script having 71 lines, then when the lsp loads it attempts to parse the file and outputs the log in the background, then in my editor, I remove a line and save, reopen the file at 70 lines and the log outputs the tree properly. I add another line again, save, reopen, then the tree ends up outputting [0, 0][0, 0] again

(I tried to embed the gif, but it was too large)
https://i.imgur.com/e0bYdmH.gifv

**Comment by ahlinc** [comment, marked as answer]
![Screenshot from 2023-04-24 19-36-48](https://user-images.githubusercontent.com/14666676/234060576-0933d7e8-8176-4e00-91c1-10f4cc4c9676.png)

I've tried your grammar on a doubled input from the gist and Tree-sitter parses it without any abruptions.
I'd suggest:
- Try to isolate your edit in a simple example outside of the complex runtime and debug it.
  - It looks like you have a bug in an edit flow implementation on your LSP and there may be many places where the bug may exist (like external scanner, etc.)
- Fix the grammar, on the screenshot you can see that it has an issue with parsing empty fn braces.
- Add tests for your grammar that would be run by `tree-sitter test`.

**Reply by MostHated** [reply]
Sorry to bring this up again, but it's going to haunt me if I don't ask. 

When you said 'It looks like you have a bug in an edit flow', did you mean that you specifically noticed one that I was not seeing, or that there may be one that was causing the initial issue?

**Reply by ahlinc** [reply]
That was just an assumption because I know on my own experience that it's not easy to implement a correct edits flow that wouldn't leak by memory or hung on CPU.

**Comment by MostHated** [comment]
Ah, shoot. That is my bad again. I had already fixed that, as well as a few others. I just had not pushed it up to the repo (just did). There were no grammar errors when I tested it. 

Thats the thing that is throwing me off, the fact that there were no grammar errors, and `tree-sitter parse` worked fine, it was only when running the parse command using the generated rust bindings that it happens.

Side note, I tried running the command that you ran, as I had never seen that particular output before, `tree-sitter p -oc1 ` but it didn't seem to work out:

```sh
zsh ❯ tree-sitter p -oc1 <(cat sample_scripts/string_test.cy{,})
error: Found argument '-o' which wasn't expected, or isn't valid in this context
```
Is that a different tool than the standard tree-sitter binary?

That said, I definitely appreciate you having taken the time to look into this. Again, my apologies for not having updated the repo. I am currently the only one who uses it, so I had not been in a huge rush yet to get things out there so far
I am going to try again to put together a standalone repro. I tried earlier, but ran into an issue trying to have tree-sitter-cyber as a dep. Due to the size of parser.c, I had to use git-lfs, not realizing that cargo doesn't seem to like that. I got side tracked trying to come up with a workaround.

**Reply by ahlinc** [reply]
> Is that a different tool than the standard tree-sitter binary?

It's my fork that isn't ready to be upstreamed yet.
You can try it with cargo: `cargo install --git https://github.com/ahlinc/tree-sitter --branch alpha` and find more info about it on [Discord](https://discord.gg/w7nTvsVJhm) server in pinned messages of general channel.

**Comment by MostHated** [comment]
Just wanted to share an update. I finally got around to getting a repro together, and naturally.. it worked fine. So I cleared everything out of that 'on_open' function and got back to basics and now it is able to handle whatever I throw at it, it seems. I copy-pasted that string test to 6,000 lines and it didn't even blink.

Thanks again for taking the time to point me in the right direction, I definitely appreciate it. :+1:
